### PR TITLE
#438 Scan jars for dalvik precompiled classes

### LINKED ISSUE #438: Include dalvik compiled classes from third party libraries

### PR DESCRIPTION
**Pull request by jperedadnr**
This PR adds support to scanning all the jars in the classpath for precompiled classes under META-INF/substrate/dalvik

These classes will be added to the gvm/apk/class folder and included, if compiling from sources, in the classpath.

Related to Keyboard support, the Keyboard Service will include precompiled classes, once https://github.com/gluonhq/attach/pull/60 is merged.

Fixes #438 and #437

### CODE REVIEW

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 1]
```diff
@@ -348,8 +353,9 @@ private void ensureApkOutputDirectoriesExist() throws IOException {
     private boolean processPrecompiledClasses(String androidJar) throws IOException, InterruptedException {
         String androidCodeLocation = "/native/android/dalvik";
 
+        copyOtherDalvikClasses();
         if (projectConfiguration.isUsePrecompiledCode()) {
-            copyPrecompiledClasses(androidCodeLocation + "/precompiled/class/");
+            copyPrecompiledClasses(androidCodeLocation + Constants.DALVIK_PRECOMPILED_CLASS);
```
Comment (eugener) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: Codacy Issue found: [Unnecessary use of fully qualified name 'Constants.DALVIK_PRECOMPILED_CLASS' due to existing static import 'com.gluonhq.substrate.Constants.DALVIK_PRECOMPILED_CLASS'](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5183430)

**Comment by eugener** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/com/gluonhq/substrate/model/ClassPath.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5183430&bid=17008393)

**Comment by eugener** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/gluonhq/substrate/model/ClassPath.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5183430&bid=17008393)

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 2]
```diff
@@ -380,12 +392,43 @@ private int compileDalvikCode(String androidSrc, String androidJar) throws IOExc
...
+        final List<File> jars = new ClassPath(projectConfiguration.getClasspath()).getJars(true);
+        String prefix = META_INF_SUBSTRATE_DALVIK + DALVIK_PRECOMPILED_CLASS;
+        for (File jar : jars) {
+            try (ZipFile zip = new ZipFile(jar)) {
+                Logger.logDebug("Scanning " + jar);
+                for (Enumeration e = zip.entries(); e.hasMoreElements(); ) {
+                    ZipEntry zipEntry = (ZipEntry) e.nextElement();
+                    String name = zipEntry.getName();
+                    if (!zipEntry.isDirectory() && name.startsWith(META_INF_SUBSTRATE_DALVIK)) {
+                        Path classPath = targetFolder.resolve(name.substring(prefix.length() + 2));
```
Comment (tiainen) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: Why +2 here?

Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: +1 gives "/com/gluonhq/helloandroid...", so resolving "/" goes to root and doesn't take into account `targetFolder`. Adding +2 removes "/" and gives "$targetFolder/com/gluonhq...".

Comment (tiainen) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: I still don't understand. The value of `prefix` is `META-INF/substrate/dalvik/precompiled/class/`. And the name of a zip entry would be `META-INF/substrate/dalvik/precompiled/class/com/hello/MyClass.class`. This means that `name.substring(prefix.length() + 2)` results in the value `m/hello/MyClass.class`?

Comment (jperedadnr) [reply src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: Should be fixed now

File: src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java [hunk 3]
```diff
@@ -322,7 +329,7 @@ private Path getApkBinPath() {
     }
 
     private Path getApkClassPath() {
```
Comment (tiainen) [inline src/main/java/com/gluonhq/substrate/target/AndroidTargetConfiguration.java]: Rename the method to `getApkClassesPath`?

**Comment by eugener** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/gluonhq/substrate/model/ClassPath.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gluonhq/substrate/pullRequest?prid=5183430&bid=17008393)

**Pull request merged**